src/core/dbus-job.c

```diff
@@ -215,6 +215,28 @@ static int send_removed_signal(sd_bus *bus, void *userdata) {
         if (!p)
                 return -ENOMEM;
 
+        /* Send the most detailed version of the signal first.
+         * Clients may rely on this order, to support versions
+         * without the most detailed signal. */
+
+        r = sd_bus_message_new_signal(
+                        bus,
+                        &m2,
+                        "/org/freedesktop/systemd1",
+                        "org.freedesktop.systemd1.Manager",
+                        "JobRemoved2");
```

Comment:
hmm, so it sucks hard that dbus doesn't allow us to extend signals with additional parameters, but I am not sure I am ready to admit defeat here, and accept that we fire two signals here...

Why not simply expose the the StopResult on the bus and query that off the unit in "systemctl" after stopping things?

Reply:
Hmm, I think I was unsure about when StopResult would be cleared.  You want to be able see StopResult after restart (stop+start).  I guess you have to define StopResult as being cleared by a successful stop, but not a successful start.

That would have been confusing if StopResult was reflected in the overall service state.  Because the overall failure is cleared on start, not stop.  I guess it's clearer though now we agreed the overall service state should not be affected by StopResult.

I'm also attracted to stuff that's more strictly non-racy :-p, but you probably have a point.  Also it will be subject to the limitation you found, of Result being defined on the specific unit type interfaces.  Neither looks like a deal-breaker, I'm mainly rationalizing why I didn't "simply" write that approach :).

Reply:
Hmm, I think the idea so far is to reset Result= + ReloadResult= each time we go through service_start(), i.e. when we begin a new iteration. And I'd assume StopResult= should be cleared then as well. (and of course in service_reset_failed(), as that's where we clear all kinds of failure states).

No, I see that this wouldn't cover the restart case, but maybe for that case it would be good enough to properly log about all this?

The raciness issue is very valid, but dbus doesn#t really make that easy. I mean, the whole concept of dbus is racy there, as sending out notifications for changes can never properly work... to be fully race-free we would always have to including the whole state in each change message... – Which makes me want to ignore this mess tbh...

Reply:
Consider the race ignored.  I'm not aware of any practical situation that makes me want to fix it.

wups, we _hadn't_ actually decided to exclude StopResult from the unit state, only the result of the job.  So there is potential confusion if you 1) clear the unit failed state on start, and leave StopResult set until the next stop.  That's why I ended up with option 2), JobRemoved2.

I'm perfectly happy to accept option 1) if the new signal is considered uglier.  I think the JobRemoved2 implementation ends up pretty simple... but if you're vaguely thinking about expanding the usage of Result, e.g. with ReloadResult, then relying purely on StopResult might make things simpler to understand.

I dunno about the other users who had this issue.  But I encountered the original issue from just experimenting with fake services.  It happened to be with `stop`, which is simpler, but I would want `restart` to feel consistent with that.  (I don't think there's a problem if we lose the error on `reset-failed` though, that part makes sense to me).

---

I don't specifically care about showing the type of error; JobRemoved2 did everything I wanted with just an extra boolean...  it could have been written as JobHadUncleanStop instead :D.  Somehow I doubt that's any more attractive to you.

I vaguely assume it's not possible to rely on a signal sent from a job object, because the job might disappear again before the client had a chance to subscribe on the job object...

It could work in principle if we emitted a PropertiesChanged signal on the service, which included the value of a new StopResult property.  But we don't have any code to send changed property values on units.

---

Option 3): There's a more generic pattern to pre-empt JobRemoved3 etc: send a dictionary of property values at removal time.  That would reduce the ugliness of the design of JobRemoved2 IMO.  The implementation will not be quite as simple, but it won't be too bad unless we went mad adding job properties.